docs/rfc/rfc-003-performance-questions.md

```diff
@@ -0,0 +1,193 @@
+
+#### Block Syncing is slow
+
+Bootstrapping a new node in a network to the height of the rest of the network
+takes longer than users would like. Block sync requires fetching all of the blocks from
+peers and placing them into the local disk for storage. A useful line of inquiry
+is understanding how quickly a perfectly tuned system _could_ fetch all of the state
+over a network so that we understand how much overhead Tendermint actually adds.
+
+The operation is likely to be _incredibly_ dependent on the environment in which
+the node is being run. The factors that will influence syncing include:
+1. Number of peers that a syncing node may fetch from.
+2. Speed of the disk that a validator is writing to.
+3. Speed of the network connection between the different peers that node is
+syncing from.
+
+We should calculate how quickly this operation _could possibly_ complete for common chains and nodes.
+To calculate how quickly this operation could possibly complete, we should assume that
+a node is reading at line-rate of the NIC and writing at the full drive speed to its
+local storage.
+
+Comparing this theoretical upper-limit to the actual sync times observed by node operators will
+give us a good point of comparison for understanding how much room for improvement there is
+within Tendermint.
+
+### Application
+
+Applications performing complex state transitions have the potential to bottleneck
+the Tendermint node.
+
+#### ABCI block delivery could cause slowdown
+
+ABCI delivers blocks in several methods: `BeginBlock`, `DeliverTx`, `EndBlock`, `Commit`.
+
+Tendermint delivers transactions one-by-one via the `DeliverTx` call. Most of the 
+transaction delivery in Tendermint occurs asynchronously and therefore appears unlikely to
+form a bottleneck in ABCI.
+
+After delivering all transactions, Tendermint then calls the `Commit` ABCI method.
+Tendermint locks all access to the mempool while `Commit` proceeds. This means that
+an application that is slow to execute all of its transactions or finalize state during
+the `Commit` method will prevent any new transactions from being added to the mempool. 
+Apps that are slow to commit will prevent consensus from proceeded to the next consensus height
+since Tendermint cannot validate validate block proposals or produce block proposals
+without the AppHash obtained from the `Commit` method.
+
+#### ABCI serialization overhead causes slowdown
+
+The most common way to run a Tendermint application is using the Cosmos-SDK.
+The Cosmos-SDK runs the ABCI application within the same process as Tendermint.
+When an application is run in the same process as Tendermint, a serialization penalty
+is not paid. This is because the local ABCI client does not serialize method calls
+and instead passes the protobuf type through through directly.
+
+Serialization and deserialization in the gRPC and socket protocol ABCI methods
+may cause slowdown. While these may cause issue, they are not part of the primary
```

Comment:
I agree this is a possible area of concern, and it's good to call it out. Have we seen serialization being a bottleneck on performance, though? While protobuf isn't the most cache-friendly encoding, it seems likely that the transaction verification & application will strongly dominate any encoding/decoding overhead unless the transactions are quite trivial.